From ba4bdc73806101dcd392b8cc63e17535b4e17ddd Mon Sep 17 00:00:00 2001
From: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Thu, 1 Dec 2022 16:36:33 -0300
Subject: [PATCH 29/55] dwarf_loader: DW_TAG_inlined_subroutine needs recoding
 by DW_AT_abstract_origin

Not just by DW_AT_type.

Signed-off-by: Arnaldo Carvalho de Melo <acme@redhat.com>
---
 dwarf_loader.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/dwarf_loader.c b/dwarf_loader.c
index 5a74035..7a6a9f8 100644
--- a/dwarf_loader.c
+++ b/dwarf_loader.c
@@ -2235,9 +2235,15 @@ static void lexblock__recode_dwarf_types(struct lexblock *tag, struct cu *cu)
 			lexblock__recode_dwarf_types(tag__lexblock(pos), cu);
 			continue;
 		case DW_TAG_inlined_subroutine:
-			dtype = dwarf_cu__find_tag_by_ref(dcu, &dpos->type);
+			if (dpos->type.off != 0)
+				dtype = dwarf_cu__find_tag_by_ref(dcu, &dpos->type);
+			else
+				dtype = dwarf_cu__find_tag_by_ref(dcu, &dpos->abstract_origin);
 			if (dtype == NULL) {
-				tag__print_type_not_found(pos);
+				if (dpos->type.off != 0)
+					tag__print_type_not_found(pos);
+				else
+					tag__print_abstract_origin_not_found(pos);
 				continue;
 			}
 			ftype__recode_dwarf_types(dtype->tag, cu);
-- 
2.40.0.rc2

